### ISSUE TITLE
Automate Registry

### ISSUE THREAD

**Issue by clsource** [joshgoebel/wren-console#7]
This small python script would be helpful to generate the modules.c file

```py
#!/usr/bin/env python3

"""
Helper tool to generate a modules.c file
containing the custom modules.
"""
import os
from pathlib import Path

base_directory = "addl_modules"

def sanitize(module):
    return module.lower().replace("_", "-")

def files():
    modules = []
    for path in Path(base_directory).rglob("*.wren"):

        """
        Allows directories. The main file
        must be named the same as the directory.
        Other files would not be included in the module.
        Useful for amalgamated files.
        Example:
        ['hello.wren', 'booger/booger.wren', 'wren_package/wren_package.wren']
        """
        directory = os.path.dirname(path)
        filename = os.path.basename(path)
        name = os.path.splitext(filename)[0]
        goodname = sanitize(name)

        if directory == base_directory:
            modules.append((goodname, name, filename, directory, path))
        else:
            if directory.endswith(name):
                modules.append((goodname, name, filename, directory, path))

    return modules

def generate(modules):
    out = """
static ModuleRegistry additionalRegistry[] =
{
    """
    for module in modules:
        goodname, name, *_ = module
        out += f"""
  NAMED_MODULE({goodname}, {name})
  END_MODULE
    """
    out += """
  SENTINEL_MODULE
};
    """
    return out


if __name__ == "__main__":
    print(generate(files()))
```

Example Output

```c
static ModuleRegistry additionalRegistry[] =
{

  NAMED_MODULE(booger, booger)
  END_MODULE

  NAMED_MODULE(hello, hello)
  END_MODULE

  NAMED_MODULE(wren-package, wren_package)
  END_MODULE

  SENTINEL_MODULE
};
```

**Comment by joshgoebel** [comment]
_First, I agree a tool like this would be awesome to have._ :-)

I'm not sure how interested we are in adding *additional* Python tooling though, unless it's slated/guaranteed to be upstreamed back into the official CLI.  IE, If we're going to work closely together (and they prefer to stick with Python) then I'm a bit more open to this - to make collaboration easier.  If it turns out our goals are less aligned and we're going to be more or less on our own then I'd much, much rather all our tooling ultimately be pure Wren - for maximum portability, fewer upfront requirements, and eating our own dog food.  

So long as the default checkout can build an initial working binary without any tooling required then a binary isn't even a prerequisite... so one can build their first binary and then use that to run the tooling to make changes for future builds. A very easy bootstrapping experience.

---

In this specific case though I'm not even sure the official CLI is interested in making additional modules like this so easy to "drop in"... so unless that's first established I absolutely think this tool should be Wren, not Python.  A lot is still up in the air regarding how the CLI and Console projects might work together in the future.

**Comment by joshgoebel** [comment]
We probably long-term want a separate `additional_modules.c` and `additional_modules.h`... so they can be easily rebuilt without mucking with the more built-in stuff in `modules.c`.

**Comment by joshgoebel** [comment]
This is exactly the kind of small utility script that *should* be trivial to write in Wren CLI/Console... and if there are missing pieces/libraries then I'd prefer we create issues for them and slowly close them out (whether here or in essentials).  IE, Python might be "ok" (for now) if there are truly good reasons this can't be currently done in pure Wren... but as soon as it could be, we should switch.

So, is there any major blockers preventing this from being Wren?

**Comment by clsource** [comment]
I wrote it in Python as a proof of concept. I think it can be done in Wren too since it just outputs text/files.
But before doing the Wren implementation I think is best to define first if the `additional_modules.c` would be used, so it's generation is automated in a single command.

Also the Wren 2 C output seems like a good candidate to Wren evolution too :)

**Comment by joshgoebel** [comment]
> I think is best to define first if the additional_modules.c would be used,

I can't think to a disadvantage of it being in a isolated file. Can you?

> Also the Wren 2 C output seems like a good candidate to Wren evolution too :)

It absolutely would.  So much so that it's one of the first things I did.  I still have a branch/PR of that somewhere... but mainline CLI didn't seem to agree with me about the utility of the build tools being in pure Wren - which is why now I'm taking a bit of a wait and see approach to see how closely we might work together in the future before touching all the existing tools.

**Comment by clsource** [comment]
I think then it would be nice to have another repo with a toolbelt of helper wren scripts.
So it easier to write such tools and not be depending on the main cli to be implemented.

They can be optional if you want. Maybe add a Makefile to ease their download and automate its calls :)

**Comment by clsource** [comment]
Another option is having a separate branch from the main one that contains all the `wren console` exclusive modifications to the official cli.

**Comment by joshgoebel** [comment]
> I think then it would be nice to have another repo with a toolbelt of helper wren scripts.

This is sort of why [wren-package](https://github.com/joshgoebel/wren-package) exists, though it was only a very minimal first pass just to get things rolling.  If someone wants to simply install libraries they should install them into `~/wren_modules/`... this can be as simple as `git checkout whatever`, which is exactly how wren-package installs packages.

If someone wants to install "utility scripts" they can just put them anywhere, in a common folder, in the system path, etc...  There is no need for "helper wren scripts" to be compiled into the CLI.  `additional_modules` exists to make it easier for me (and others) to build custom CLIs (with embedded code).  However, this is not something the *average* scripting user should need to do.

Related: https://github.com/wren-lang/wren-cli/issues/78

---

Perhaps I'm just not sure exactly what you mean by "helper wren scripts"...

**Comment by clsource** [comment]
Mainly I was refering to utilities and scripts related to `wren console` development itself. 
Like the additional_modules generator or the wren2c script.

Other scripts in userland is best to use wren-package or related tooling.

**Comment by joshgoebel** [comment]
> Another option is...

Eventually if CLI wants to work more closely (sharing between CLI and Console) we'll have to discuss/discover how that works best.  But it seems that needs to be decided first before getting into any specific details of how to manage that.  Console is a direct fork of CLI with shared history so many things (that we still share in common) should be *fairly* easy to merge/cherry-pick over even without any of the changes you propose.

> Mainly I was refering to utilities and scripts related to wren console development itself. Like the additional_modules generator or the wren2c script.

I don't see any advantage to that living in a separate repository. Those would be core parts of wren-console, and hence live in this repo.

**Comment by clsource** [comment]
Yes right now it easy to patch wren console and pick wren-cli improvements.
But is a one way only.

I think wren-cli maybe in the future could benefit from the development of wren-console
in the form of specific PRs that maybe accepted (or not).

So paving a way to a bidirectional file and branch structure can be a good idea.

**Comment by joshgoebel** [comment]
> But is a one way only.

Not at all. It's quite easy for someone working on wren-cli to add wren-console as a remote and cherry-pick commits or generate patches, etc... in some cases it'll work magically (thanks to our shared history) - in other cases it may require resolving slight differences (if the codebases drift further apart), etc.

> I think wren-cli maybe in the future could benefit from the development of wren-console

I'd like to hope so. :-)  And vise-versa.

> So paving a way to a bidirectional file and branch structure can be a good idea.

Branches names really don't matter much practically speaking, they are just pointers to SHAs... No specific branch names are needed (AFAIK) for two repos in this situation to share things back and further...

Directory layout is a slightly more interesting consideration, but I'm not super worried about that until it proves to be a problem....

**Comment by clsource** [comment]
Fair enough.
The branch names were only conventions to keep things tidy :).

**Comment by joshgoebel** [comment]
Anyways things like this I prefer to deal with in actual usage.  If Wren proper decides to start pulling some of my work back into CLI, and we run into problems with the practicality of doing so, then at the time we find the problems we'll look for the solutions.  When the most recent `chdir` patch is merged I might try a quick PR against CLI, and I expect it to be trivial.

**Comment by clsource** [comment]
Maybe this script would be best to wren essentials https://github.com/joshgoebel/wren-essentials

**Comment by joshgoebel** [comment]
I think if people want to compile a CLI with 'custom stuff' the right place to do it is at the CLI level, not wren-essentials.... so the tool belongs here - just we need to sort out whether it should be Python or Wren.

Did you not want to just port it to Wren and get it added here?

**Comment by clsource** [comment]
I just thought that since wren-console must adhere to stricter file and directory structures and other constraints to be wren-cli compatible that moving this tool to wren-essentials can be a smoother move.

Both would require a similar tool to generate the C files.

Maybe this tool can be on its own and be used as a dep?

**Comment by joshgoebel** [comment]
> Both would require a similar tool to generate the C files.

Well, maybe, but not necessarily.  I'd love to auto-generated all the C stubs from source alone eventually (including `foreign` signatures, etc) but that's not what this does.  This just takes `addl_modules` and builds a tiny C stub for them.  That's needed for CLI but not (yet) needed for essentials which doesn't yet have an idea of just vendoring other libraries wholesale (and maybe never will).

If we were doing the same thing in both places then this might indeed make sense to break out into a Wren -> C header type of tool - though surely such a tool would have uses on it's own if it worked well.

Another thought is once we were generating it programmatically all the magic DEFINES in the C preprocessor would no longer be needed either and we could just spew out ugly C code, because we weren't maintaining it by hand.  I"m not sure if that would be better or worse though without trying it. (as then it just pushed that complexity into the Wren -> C tool.

---

The tooling that IS shared right now is the `wren_to_c_string.py` tooling, but I don't think that justifies a repo of it's own yet.  My version is already more advanced than the CLI one to allow me to put source into different (or multiple) modules if I so choose.

**Comment by clsource** [comment]
In that case maybe creating a new directory in root as `tools` maybe?
so wren only tools can be develop to help wren-console devs?
since util would be python only with scripts that cannot be made with wren because of different reasons?
